### PR TITLE
Random passphrase - clientside

### PR DESCRIPTION
**Pull request by zapu**

### CODE REVIEW

File: shared/actions/users.js [hunk 1]
```diff
@@ -33,9 +36,24 @@ const updateProofState = (state, action) => {
   }
 }
 
+const setupEngineListeners = () => {
+  engine().setIncomingCallMap({
+    'keybase.1.NotifyUsers.passwordChanged': () =>
```
Comment (buoyad) [inline shared/actions/users.js]: This could go in `actions/settings` so it's all in one place

File: shared/actions/settings.js [hunk 2]
```diff
@@ -414,6 +414,17 @@ const unfurlSettingsSaved = (state, action) =>
       })
     )
 
+const loadHasRandomPW = state => {
```
Comment (buoyad) [inline shared/actions/settings.js]: 🐑 you could shorten this to `state => state.settings.passphrase.randomPW === null ? callRPCStuff() : null`

File: shared/settings/index.js [hunk 3]
```diff
@@ -4,36 +4,56 @@ import * as ConfigGen from '../actions/config-gen'
 import * as Types from '../constants/types/settings'
 import * as RouteTreeGen from '../actions/route-tree-gen'
 import SettingsContainer from './render'
-import {connect} from '../util/container'
+import {createLoadHasRandomPw} from '../actions/settings-gen'
```
Comment (buoyad) [inline shared/settings/index.js]: `import * as SettingsGen`

File: shared/settings/index.js [hunk 4]
```diff
@@ -4,36 +4,56 @@ import * as ConfigGen from '../actions/config-gen'
...
   onLogout: dispatchProps.onLogout,
   onTabChange: dispatchProps.onTabChange,
   selectedTab: stateProps.selectedTab,
 })
 
-export default connect<OwnProps, _, _, _, _>(
-  mapStateToProps,
-  mapDispatchToProps,
-  mergeProps
+// export default connect<OwnProps, _, _, _, _>(
```
Comment (buoyad) [inline shared/settings/index.js]: 🔥

File: shared/settings/index.js [hunk 5]
```diff
@@ -4,36 +4,56 @@ import * as ConfigGen from '../actions/config-gen'
...
+
+export default compose(
+  connect<OwnProps, _, _, _, _>(
+    mapStateToProps,
+    mapDispatchToProps,
+    mergeProps
+  ),
+  lifecycle({
+    componentDidMount() {
+      console.log('::::::: componentDidMount for settings/index')
```
Comment (buoyad) [inline shared/settings/index.js]: 🔥

File: shared/settings/landing/index.desktop.js [hunk 6]
```diff
@@ -294,26 +294,47 @@ function AccountPassphrase({onChangePassphrase}: {onChangePassphrase: () => void
...
-      <AccountPassphrase onChangePassphrase={onChangePassphrase} />
-      <Divider />
-      <Checkbox
-        checked={rememberPassphrase}
-        label="Remember my passphrase"
-        onCheck={onChangeRememberPassphrase}
-        style={{paddingTop: globalMargins.small}}
-      />
+      <Passphrase onChangePassphrase={onChangePassphrase} />
+      {!hasRandomPW && <Divider />}
```
Comment (buoyad) [inline shared/settings/landing/index.desktop.js]: You can do this and the `Checkbox` together by surrounding them with `<> ... </>`
```
{!hasRandomPW && (<><Thing1 /><Thing2 /></>)}`

Comment (buoyad) [reply shared/settings/landing/index.desktop.js]: Also, do you care whether it's `false` or `null` here?

File: shared/settings/nav/index.native.js [hunk 7]
```diff
@@ -74,7 +74,8 @@ function SettingsNav(props: Props) {
             },
             {
               onClick: () => props.onTabChange(Constants.passphraseTab),
-              text: 'Change passphrase',
+              text: !props.hasRandomPW ? 'Change passphrase' : 'Set a passphrase',
+              textColor: !props.hasRandomPW ? undefined : globalColors.red,
```
Comment (buoyad) [inline shared/settings/nav/index.native.js]: Easier to read if you test for the positive here

File: shared/settings/index.js (hunk 5 above)
Comment (zapu) [reply shared/settings/index.js]: whoooops

File: shared/settings/landing/index.desktop.js (hunk 6 above)
Comment (zapu) [reply shared/settings/landing/index.desktop.js]: if it's `null` it should assume that user has passphrase, so none of the special NOPW treatment should happen. I think it's fine if it refreshes while user is in there.

**Comment by zapu** [comment]
@buoyad can you take another look? this might be ready, serverside changes have been merged and will be deployed shortly. I'm still doing some testing.

File: shared/settings/advanced/container.js [hunk 8]
```diff
@@ -15,13 +16,15 @@ import {connect, lifecycle} from '../../util/container'
 
 type OwnProps = {||}
 const mapStateToProps = state => ({
+  hasRandomPW: state.settings.passphrase.randomPW,
```
Comment (buoyad) [inline shared/settings/advanced/container.js]: I think this should be `!!state.settings.passphrase.randomPW`, flow is letting it pass because the `lifecycle` below is losing the component's type.
